deepinv/transform/base.py

```diff
@@ -1,30 +1,301 @@
+                    },
+                )
+                for i in range(len(x))
+            ]
+        )
+
+    def identity(self, x: torch.Tensor, average: bool = False) -> torch.Tensor:
+        """Sanity check function that should do nothing.
+
+        This performs forward and inverse transform, which results in the exact original, down to interpolation and padding effects.
+
+        Interpolation and padding effects will be visible in non-pixelwise transformations, such as arbitrary rotation, scale or projective transformation.
+
+        :param torch.Tensor x: input image
+        :param bool average: average over ``n_trans`` transformed versions to get same number as output images as input images. No effect when ``n_trans=1``.
+        :return torch.Tensor: :math:`T_g^{-1}T_g x=x`
+        """
+        return self.symmetrize(f=lambda _x: _x, average=average)(x)
+
+    def iterate_params(self, params):
+        negs = [getattr(p, "_neg", None) for p in params.values()]
+        param_lists = [p.tolist() for p in params.values()]
+
+        return [
+            {
+                key: (
+                    torch.tensor([comb[i]])
+                    if negs[i] is None
+                    else TransformParam([comb[i]], neg=negs[i])
+                )
+                for i, key in enumerate(params.keys())
+            }
+            for comb in list(product(*param_lists))
+        ]
+
+    def symmetrize(
+        self, f: Callable[[torch.Tensor, Any], torch.Tensor], average: bool = False
+    ) -> Callable[[torch.Tensor, Any], torch.Tensor]:
+        r"""
+        Symmetrise a function with a transform and its inverse.
+
+        Given a function :math:`f(\cdot):X\rightarrow X` and a transform :math:`T_g`, returns the group averaged function  :math:`\sum_{i=1}^N T_{g_i}^{-1} f(T_{g_i} \cdot)` where :math:`N` is the number of random transformations.
+
+        This is useful for e.g. Reynolds averaging a function over a group. Set ``average=True`` to average over n_trans.
+        For example, use ``Rotate(n_trans=4, positive=True, multiples=90).symmetrize(f)`` to symmetrize f over the entire group.
+
+        :param Callable[[torch.Tensor, Any], torch.Tensor] f: function acting on tensors.
+        :param bool average: monte carlo average over all random transformations (in range ``n_trans``) when symmetrising to get same number of output images as input images. No effect when ``n_trans=1``.
+        :return Callable[[torch.Tensor, Any], torch.Tensor]: decorated function.
+        """
+
+        def symmetrized(x, *args, **kwargs):
+            params = self.get_params(x)
+            if self.constant_shape:
+                # Collect over n_trans
+                xt = self.inverse(
+                    f(self.transform(x, **params), *args, **kwargs),
+                    batchwise=False,
+                    **params,
+                )
+                return xt.reshape(-1, *x.shape).mean(axis=0) if average else xt
```

Comment:
the current implementation uses |G| extra memory right?
can we add an option that uses |G| extra compute also?